plugin/completion.py

```diff
@@ -105,62 +56,52 @@ def format_completion(item: dict, restore_lines: RestoreLines) -> sublime.Comple
         kind = tuple(list_kind)  # type: ignore
 
+    item["change_id"] = change_id
+
     return sublime.CompletionItem.command_completion(
         trigger=item["label"],
         command="lsp_select_completion_item",
-        args={
-            "item": item,
-            "restore_lines_dict": restore_lines.to_dict()
-        },
+        args=item,
         annotation=item.get('detail') or "",
         kind=kind
     )
 
 
 class LspSelectCompletionItemCommand(sublime_plugin.TextCommand):
-    def run(self, edit: sublime.Edit, item: Any, restore_lines_dict: dict) -> None:
-        insert_text_format = item.get("insertTextFormat")
+    """
+    This command must handle four different kinds of LSP completion items:
+
+    1) plaintext + insertText   (e.g. pyls)
+    2) plaintext + textEdit     (e.g. intelephense)
+    3) snippet   + insertText   (???)
+    4) snippet   + textEdit     (e.g. clangd, intelephense)
+
+    For cases (3) and (4) we are forced to use the "insert_snippet" command.
+    """
 
+    def run(self, edit: sublime.Edit, **item: Any) -> None:
+        # Is it a textEdit or an insertText?
         text_edit = item.get('textEdit')
         if text_edit:
-            # restore the lines
-            # so we don't have to calculate the offset for the textEdit range
-            restore_lines = RestoreLines.from_dict(restore_lines_dict)
-            restore_lines.restore_lines(edit, self.view)
-
-            new_text = text_edit.get('newText')
-
-            range = Range.from_lsp(text_edit['range'])
-            edit_region = range_to_region(range, self.view)
-
-            # calculate offset by comparing cursor position with edit_region.begin.
-            # by applying the offset to all selections
-            # the TextEdit becomes valid for all selections
-            cursor = self.view.sel()[0].begin()  # type: int
-
-            offset_start = cursor - edit_region.begin()
-            offset_length = edit_region.end() - edit_region.begin()
-
-            # erease regions from bottom to top
-            for sel in reversed(self.view.sel()):
-                begin = sel.begin() - offset_start
-                end = begin + offset_length
-                r = sublime.Region(begin, end)
-                self.view.erase(edit, r)
-
-            if insert_text_format == InsertTextFormat.Snippet:
-                self.view.run_command("insert_snippet", {"contents": new_text})
-            else:
-                # insert text from bottom to top
-                for sel in reversed(self.view.sel()):
-                    self.view.insert(edit, sel.begin(), new_text)
+            new_text = text_edit['newText']
+            # this region was valid a few view.change_count() moments back ...
+            edit_region = range_to_region(Range.from_lsp(text_edit['range']), self.view)
+            # ... but this brings it to the present.
+            edit_region = self.view.transform_region_from(edit_region, item["change_id"])
```

Comment:
I am still confused what region should be passed to transform_region_from, 
but I tested this change and it works. You also deleted restore lines class... All I can say is nice job! :)

Reply:
From what I can tell, `transform_region_from` expects a region from the past.
It uses the change_id and the internal undo-stack of ST to bring the region to the present.

I noticed that in order to get this "region from the past", we need to translate the "LSP range from the past" into a "sublime region from the past", but in order to do so we need to use `range_from_region`, which uses `point_to_offset`, which uses `view.text_point_utf16`, which doesn't use the change_id. So I'm not 100% this will always work.